ONSdigital/eq-questionnaire-launcher
# Integrate sds with launcher

### What is the context of this PR?
This PR is the launcher component of the ticket to Integrate SDS with Launcher. When a schema is selected, the `survey_id` of that schema is retrieved and returned along with the metadata so that if the metadata also contains `survey_id`, then it will be pre-populated with the `survey_id` of the survey (to make life easier for devs)

The metadata `survey_id` can be changed to anything, and when both `survey_id` and `period_id` are populated fields in the metadata, and when either of them change, a new call is made to the mock endpoint in the [corresponding runner PR](https://github.com/ONSdigital/eq-questionnaire-runner/pull/1147) to fetch the SDS datasets and use them to fill the dropdown list.

### Changes
Adjusted `launch.go` to replace the `metadataHandler` with a `surveyDataHandler`. This is because we need the `survey_id` from the schema, to pre-populate any `survey_id` that might be found in metadata, and it was not enough to return the metadata alone to render the page.

Additionally in `launch.go` added a similar `supplementaryDataHandler` to intercept any requests to the `supplementary-data` url in the front-end and make the request to the mock end point with proper handling

The actual fetching of supplementary datasets is done in `surveys.go` in a similar style to other fetches and using the `DatasetMetadata` struct to model the response, as per the sds api documentation

`launch.html` has been refactored to improve the way fetches were being done, make reusable components for the various input fields and whenever the schema changes and metadata is populated, check for a survey id and period id in the metadata and fetch supplementary datasets if both can be found. 

If supplementary datasets are found, the supplementary data div is displayed and populated with the selected dataset

### How to review

1. Clone and run the [launcher PR](https://github.com/ONSdigital/eq-questionnaire-launcher/pull/49) with `go run launch.go`
2. Run the supporting services for runner with `make dev-compose-up` as normal (delete the launcher container if that is built despite launcher already being run) 
3. Launch the mock SDS from a terminal in runner`python -m scripts.mock_sds_endpoint`
4. Launch runner as normal.
5. Open launcher in chrome
6. Select V2 Launch Pattern and the schema `test_supplementary_data`
7. Survey Metadata should load as normal, with survey id 123 and a new sds_dataset_id dropdown with only one option. 
8. The first and only SDS dataset metadata should have loaded and displayed below.
9. The survey should be launchable.
10. Open launcher again
11. Select V2 Launch Pattern and the schema `test_supplementary_data_with_multiple_datasets`
12. Survey Metadata should load as normal, with survey id 456 and 2 `sds_dataset_ids` in the `sds_dataset_id` dropdown. The first one will auto-select. 
13. Swap between the given options and the Dataset Metadata section should update to reflect the metadata. 
14. Launch the survey with both options to verify it launches correctly. 
15. Open launcher again
16. Select V2 Launch Pattern and either `test_supplementary_data` or `test_supplementary_data_with_multiple_datasets`.
17. Enter a random number for the survey id in the survey metadata section.
18. The Dataset Metadata section should disappear and the `sds_dataset_id` dropdown should clear of options, and the Launch Survey button should not be available, as the survey requires an `sds_dataset_id` value.  
19. In step 17, using the id of the schema you did not load, will load the available datasets for that survey but it will be an invalid dataset for the selected survey and thus crash on launch. 

### How to review

- Read through the trello card and check that the requirements have been met, and that the changes are suitable
- As this is a dev tool, consider if this implementation offers the most helpful functionality it can for working with pre-pop tickets
